### PR TITLE
Make our shock spectra use the derivations from ISO 18431-4

### LINKED ISSUE #132: Update PVSS to match ISO standard
Modify what we have to use the ISO standard and add in the two additional spectra

### PR DESCRIPTION
**Pull request by CFlaniganMide**
This changes the coefficient calculations to use ISO 18431-4, and adds a few more shock types to the mix.  There's still a few small changes to add the new ones to `shock_spectrum` and a few more tweaks.  

As a note, I've checked and the coefficients coming out of scipy using a FOH method seem to line up almost perfectly.

closes #132

### CODE REVIEW

**Comment by CrepeGoat** [comment]
werrrrrd I'm very excited to see this happening :D keep it up!

> As a note, I've checked and the coefficients coming out of scipy using a FOH method seem to line up almost perfectly.

This *sounds* like a good thing, but I'm skeptical; I feel like it *could* just be a coincidence for the parameters you tested, or something like that. A good note, for sure, but I take it with a grain of salt.

If we *do* want to check if that holds more generally, it'd be a good application for a property (`hypothesis`) test.

**Comment by CFlaniganMide** [comment]
> werrrrrd I'm very excited to see this happening :D keep it up!
> 
> > As a note, I've checked and the coefficients coming out of scipy using a FOH method seem to line up almost perfectly.
> 
> This _sounds_ like a good thing, but I'm skeptical; I feel like it _could_ just be a coincidence for the parameters you tested, or something like that. A good note, for sure, but I take it with a grain of salt.
> 
> If we _do_ want to check if that holds more generally, it'd be a good application for a property (`hypothesis`) test.

After looking further into it, the ramp-invariant and FOH methods are the same thing.  I have no idea why basically nowhere mentions this.  But, the key difference is that I expect the ISO document has hand-derived equations and using scipy does something in state space I didn't linger on too much.  In theory, they should come out to the same answers, but naturally there's going to be some differences as a result of using A) a different method and B) floating point error.

**Comment by CrepeGoat** [comment]
> After looking further into it, the ramp-invariant and FOH methods are the same thing. 

interesting! can you compile your research on that somewhere (enough to at least show the equivalence)? I'd like to read through it.

And actually (back when I was still thinking I'd implement this myself) I was really hoping that one of the SciPy methods would align with the direct equations, because it means you can use one method to validate the other in (property) tests. I actually put in a bit of work on that a few months ago; if you're interested you can take a look at [my old branch for it](/MideTechnology/endaq-python/tree/feature/shock-ISO-18431-4) that I've rebased onto development.

> In theory, they should come out to the same answers, but naturally there's going to be some differences as a result of using A) a different method and B) floating point error.

Hopefully the divergence is small enough that for most of the input variable space the two are roughly the same. If not the property tests won't work very well :\ (but if they don't work, we shouldn't be making that guarantee anyway. who knows.)

**Comment by CFlaniganMide** [comment]
> > After looking further into it, the ramp-invariant and FOH methods are the same thing.
> 
> interesting! can you compile your research on that somewhere (enough to at least show the equivalence)? I'd like to read through it.

[This paper](https://www.sciencedirect.com/science/article/pii/S1474667017506485) is the best example I have describing it directly.  Otherwise I've been comparing the math between different sources that use different names.  

[This paper from VibrationData](http://www.vibrationdata.com/ramp_invariant/ramp_invariant_base.pdf) refers to the ramp-invariant method, check equation 13 at the bottom of page 3.

[Wikipedia has a reasonable definition of First Order Hold](https://en.wikipedia.org/wiki/First-order_hold) which I've seen cited (for figures, not the math) twice, so I'm assuming enough academics have looked at it that it's reliable.  

The gist is, in the laplace domain, multiply by a ramp function `1/(s^2)` and in the Z domain, multiply by `((1 - z^-1)/T)^2`.  This is, more specifically, the delayed FOH, which is equivalent to all the definitions of the ramp-invariant method I've seen.

**Comment by CFlaniganMide** [comment]
Clarifying the above slightly, and using [this](https://lpsa.swarthmore.edu/LaplaceZTable/LaplaceZFuncTable.html) table as reference, a simpler way to put is is that you basically multiply by the ___-invariant in the s domain, and divide by that in the z-domain.

File: endaq/calc/shock.py [hunk 1]
```diff
@@ -17,29 +17,186 @@
...
+    absolute acceleration according to ISO 18431-4
+
+    :param omega: the natural frequency of the system
+    :param Q: the quality factor of the system
+    :param T: the time step in seconds
+    :return: the coefficients of the Z-domain transfer function b, a
+
+    .. seealso::
+
+        - `ISO 18431-4 Mechanical vibration and shock — Signal processing — Part 4: Shock-response spectrum analysis`__
```
Comment (CrepeGoat) [inline endaq/calc/shock.py:32]:
```suggestion
        - `ISO 18431-4 Mechanical vibration and shock — Signal processing — Part 4: Shock-response spectrum analysis`
```

File: endaq/calc/shock.py [hunk 2]
```diff
@@ -17,29 +17,186 @@
...
+    .. seealso::
+
+        - `An Introduction To The Shock Response Spectrum, Tom Irvine, 9 July 2012 <http://www.vibrationdata.com/tutorials2/srs_intr.pdf>`_
+        - `SciPy transfer functions <https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.TransferFunction.html>`_
+          Documentation for the transfer function class used to characterize the
+          relative displacement calculation.
+        - `SciPy biquad filter <https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.lfilter.html>`__
+          Documentation for the biquad function used to implement the transfer
+          function.
+        - `ISO 18431-4 Mechanical vibration and shock — Signal processing — Part 4: Shock-response spectrum analysis`__
```
Comment (CrepeGoat) [inline endaq/calc/shock.py:75]:
```suggestion
        - `ISO 18431-4 Mechanical vibration and shock — Signal processing — Part 4: Shock-response spectrum analysis`
```

File: endaq/calc/shock.py [hunk 3]
```diff
@@ -17,29 +17,186 @@
...
+    relative velocity according to ISO 18431-4
+
+    :param omega: the natural frequency of the system
+    :param Q: the quality factor of the system
+    :param T: the time step in seconds
+    :return: the coefficients of the Z-domain transfer function b, a
+
+    .. seealso::
+
+        - `ISO 18431-4 Mechanical vibration and shock — Signal processing — Part 4: Shock-response spectrum analysis`__
```
Comment (CrepeGoat) [inline endaq/calc/shock.py:103]:
```suggestion
        - `ISO 18431-4 Mechanical vibration and shock — Signal processing — Part 4: Shock-response spectrum analysis`
```

File: endaq/calc/shock.py [hunk 4]
```diff
@@ -17,29 +17,186 @@
...
     .. seealso::
 
         - `Pseudo Velocity Shock Spectrum Rules For Analysis Of Mechanical Shock, Howard A. Gaberson <https://info.endaq.com/hubfs/pvsrs_rules.pdf>`_
         - `SciPy transfer functions <https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.TransferFunction.html>`_
           Documentation for the transfer function class used to characterize the
           relative displacement calculation.
+        - `SciPy biquad filter <https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.lfilter.html>`__
+          Documentation for the biquad function used to implement the transfer
+          function.
+        - `ISO 18431-4 Mechanical vibration and shock — Signal processing — Part 4: Shock-response spectrum analysis`__
```
Comment (CrepeGoat) [inline endaq/calc/shock.py:148]:
```suggestion
        - `ISO 18431-4 Mechanical vibration and shock — Signal processing — Part 4: Shock-response spectrum analysis`
```

File: endaq/calc/shock.py [hunk 5]
```diff
@@ -17,29 +17,186 @@
...
-            [1, 2 * damp * omega, omega ** 2],
-        ).to_discrete(dt=dt)
+    :param omega: the natural frequency of the system
+    :param Q: the quality factor of the system
+    :param T: the time step in seconds
+    :return: the coefficients of the Z-domain transfer function b, a
+
+    .. seealso::
+
+        - `ISO 18431-4 Mechanical vibration and shock — Signal processing — Part 4: Shock-response spectrum analysis`__
```
Comment (CrepeGoat) [inline endaq/calc/shock.py:176]:
```suggestion
        - `ISO 18431-4 Mechanical vibration and shock — Signal processing — Part 4: Shock-response spectrum analysis`
```

File: endaq/calc/shock.py [hunk 6]
```diff
@@ -62,75 +219,170 @@ def rel_displ(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFr
         - `SciPy transfer functions <https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.TransferFunction.html>`_
           Documentation for the transfer function class used to characterize the
           relative displacement calculation.
-        - `SciPy biquad filter <https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.lfilter.html>`_
+        - `SciPy biquad filter <https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.lfilter.html>`__
           Documentation for the biquad function used to implement the transfer
           function.
+        - `ISO 18431-4 Mechanical vibration and shock — Signal processing — Part 4: Shock-response spectrum analysis`__
```
Comment (CrepeGoat) [inline endaq/calc/shock.py:225]:
```suggestion
        - `ISO 18431-4 Mechanical vibration and shock — Signal processing — Part 4: Shock-response spectrum analysis`
```

File: endaq/calc/shock.py [hunk 7]
```diff
@@ -62,75 +219,170 @@ def rel_displ(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFr
...
+    .. seealso::
+
+        - `Pseudo Velocity Shock Spectrum Rules For Analysis Of Mechanical Shock, Howard A. Gaberson <https://info.endaq.com/hubfs/pvsrs_rules.pdf>`_
         - `SciPy transfer functions <https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.TransferFunction.html>`_
           Documentation for the transfer function class used to characterize the
           relative displacement calculation.
+        - `SciPy biquad filter <https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.lfilter.html>`__
+          Documentation for the biquad function used to implement the transfer
+          function.
+        - `ISO 18431-4 Mechanical vibration and shock — Signal processing — Part 4: Shock-response spectrum analysis`__
```
Comment (CrepeGoat) [inline endaq/calc/shock.py:298]:
```suggestion
        - `ISO 18431-4 Mechanical vibration and shock — Signal processing — Part 4: Shock-response spectrum analysis`
```

File: endaq/calc/shock.py [hunk 8]
```diff
@@ -62,75 +219,170 @@ def rel_displ(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFr
...
-        - `An Introduction To The Shock Response Spectrum, Tom Irvine, 9 July 2012 <http://www.vibrationdata.com/tutorials2/srs_intr.pdf>`_
+        - `Pseudo Velocity Shock Spectrum Rules For Analysis Of Mechanical Shock, Howard A. Gaberson <https://info.endaq.com/hubfs/pvsrs_rules.pdf>`_
         - `SciPy transfer functions <https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.TransferFunction.html>`_
           Documentation for the transfer function class used to characterize the
           relative displacement calculation.
-        - `SciPy biquad filter <https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.lfilter.html>`_
+        - `SciPy biquad filter <https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.lfilter.html>`__
           Documentation for the biquad function used to implement the transfer
           function.
+        - `ISO 18431-4 Mechanical vibration and shock — Signal processing — Part 4: Shock-response spectrum analysis`__
```
Comment (CrepeGoat) [inline endaq/calc/shock.py:372]:
```suggestion
        - `ISO 18431-4 Mechanical vibration and shock — Signal processing — Part 4: Shock-response spectrum analysis`
```

File: endaq/calc/shock.py [hunk 9]
```diff
@@ -62,75 +219,170 @@ def rel_displ(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFr
...
-    The "absolute acceleration" follows the transfer function:
-        H(s) = L{x"(t)}(s) / L{y"(t)}(s) = X(s)/Y(s)
+    :param omega: the natural frequency of the system
+    :param Q: the quality factor of the system
+    :param T: the time step in seconds
+    :return: the coefficients of the Z-domain transfer function b, a
+
+    .. seealso::
+
+        - `ISO 18431-4 Mechanical vibration and shock — Signal processing — Part 4: Shock-response spectrum analysis`__
```
Comment (CrepeGoat) [inline endaq/calc/shock.py:327]:
```suggestion
        - `ISO 18431-4 Mechanical vibration and shock — Signal processing — Part 4: Shock-response spectrum analysis`
```

**Review by CrepeGoat — CHANGES_REQUESTED** [review]
I still owe you a substantive review, but the following changes should get the docs tests to pass.

File: endaq/calc/shock.py [hunk 10]
```diff
@@ -17,29 +17,186 @@
 from endaq.calc import utils
 
 
-def _rel_displ_transfer_func(
-    omega: float, damp: float = 0.0, dt: float = 1.0
-) -> scipy.signal.ltisys.TransferFunctionDiscrete:
+def _abs_accel_coeffs(omega, Q, T):
```
Comment (CrepeGoat) [inline endaq/calc/shock.py:20]: abbreviated form: as-is 👍
non-abbreviated form:
```suggestion
def _absolute_acceleration_coeffs(omega, Q, T):
```

File: endaq/calc/shock.py [hunk 11]
```diff
@@ -17,29 +17,186 @@
...
+    a = (
+        1.,
+        -2.*np.exp(-A)*np.cos(B),
+        np.exp(-2.*A),
+        )
+
+    return b, a
+
+
+def abs_accel(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFrame:
```
Comment (CrepeGoat) [inline endaq/calc/shock.py:52]: abbreviated form: as-is 👍
non-abbreviated form:
```suggestion
def absolute_acceleration(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFrame:
```

File: endaq/calc/shock.py [hunk 12]
```diff
@@ -17,29 +17,186 @@
...
+        functools.partial(
+                scipy.signal.lfilter,
+                *_abs_accel_coeffs(omega, Q, T),
+                axis=0,
+                ),
+        raw=True,
+        )
+
+
+def _rel_velocity_coeffs(omega, Q, T):
```
Comment (CrepeGoat) [inline endaq/calc/shock.py:91]: abbreviated form:
```suggestion
def _rel_vel_coeffs(omega, Q, T):
```

non-abbreviated form:
```suggestion
def _relative_velocity_coeffs(omega, Q, T):
```

File: endaq/calc/shock.py [hunk 13]
```diff
@@ -17,29 +17,186 @@
...
+    a = (
+        1.,
+        -2.*np.exp(-A)*np.cos(B),
+        np.exp(-2.*A),
+        )
+
+    return b, a
+
+
+def rel_velocity(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFrame:
```
Comment (CrepeGoat) [inline endaq/calc/shock.py:125]: abbreviated form:
```suggestion
def rel_vel(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFrame:
```

non-abbreviated form:
```suggestion
def relative_velocity(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFrame:
```

File: endaq/calc/shock.py [hunk 14]
```diff
@@ -17,29 +17,186 @@
...
+        functools.partial(
+                scipy.signal.lfilter,
+                *_rel_velocity_coeffs(omega, Q, T),
+                axis=0,
+                ),
+        raw=True,
+        )
+
+
+def _rel_displ_coeffs(omega, Q, T):
```
Comment (CrepeGoat) [inline endaq/calc/shock.py:164]: abbreviated form: as-is 👍

non-abbreviated form:
```suggestion
def _relative_displacement_coeffs(omega, Q, T):
```

File: endaq/calc/shock.py [hunk 15]
```diff
@@ -17,29 +17,186 @@
...
+    a = (
+        1.,
+        -2.*np.exp(-A)*np.cos(B),
+        np.exp(-2.*A),
+        )
+
+    return b, a
 
 
 def rel_displ(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFrame:
```
Comment (CrepeGoat) [inline endaq/calc/shock.py:202]: abbreviated form: as-is 👍

non-abbreviated form:
```suggestion
def relative_displacement(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFrame:
```

File: endaq/calc/shock.py [hunk 16]
```diff
@@ -62,75 +219,170 @@ def rel_displ(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFr
...
+    a = (
+        1.,
+        -2.*np.exp(-A)*np.cos(B),
+        np.exp(-2.*A),
+        )
+
+    return b, a
+
+
+def pseudo_velocity(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFrame:
```
Comment (CrepeGoat) [inline endaq/calc/shock.py:275]: abbreviated form:
```suggestion
def pseudo_vel(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFrame:
```

non-abbreviated form: as-is 👍

File: endaq/calc/shock.py [hunk 17]
```diff
@@ -62,75 +219,170 @@ def rel_displ(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFr
...
+                ),
         raw=True,
-    )
+        )
 
 
-def _abs_accel_transfer_func(
-    omega: float, damp: float = 0.0, dt: float = 1.0
-) -> scipy.signal.ltisys.TransferFunctionDiscrete:
+def _pseudo_velocity_coeffs(omega, Q, T):
```
Comment (CrepeGoat) [inline endaq/calc/shock.py:241]: abbreviated form:
```suggestion
def _pseudo_vel_coeffs(omega, Q, T):
```

non-abbreviated form: as-is 👍

File: endaq/calc/shock.py [hunk 18]
```diff
@@ -62,75 +219,170 @@ def rel_displ(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFr
...
+                scipy.signal.lfilter,
+                *_pseudo_velocity_coeffs(omega, Q, T),
+                axis=0,
+                ),
+        raw=True,
+        )
 
 
-def abs_accel(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFrame:
+def _relative_disp_static_coeffs(omega, Q, T):
```
Comment (CrepeGoat) [inline endaq/calc/shock.py:314]: This one I think should also have the word "acceleration" in there, but that's just my preference.

abbreviated form:
```suggestion
def _rel_displ_static_coeffs(omega, Q, T):
```

abbreviated form w/ "acceleration":
```suggestion
def _rel_displ_static_accel_coeffs(omega, Q, T):
```

non-abbreviated form:
```suggestion
def _relative_displacement_static_coeffs(omega, Q, T):
```

non-abbreviated form w/ "acceleration":
```suggestion
def _relative_displacement_static_acceleration_coeffs(omega, Q, T):
```

File: endaq/calc/shock.py [hunk 19]
```diff
@@ -62,75 +219,170 @@ def rel_displ(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFr
...
+    a = (
+        1.,
+        -2.*np.exp(-A)*np.cos(B),
+        np.exp(-2.*A),
+        )
+
+    return b, a
+
+
+def relative_disp_static(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFrame:
```
Comment (CrepeGoat) [inline endaq/calc/shock.py:349]: abbreviated form:
```suggestion
def rel_displ_static(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFrame:
```
abbreviated form w/ "acceleration":
```suggestion
def rel_displ_static_accel(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFrame:
```

non-abbreviated form:
```suggestion
def relative_displacement_static(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFrame:
```

non-abbreviated form w/ "acceleration":
```suggestion
def relative_displacement_static_acceleration(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFrame:
```

**Review by CrepeGoat — CHANGES_REQUESTED** [review]
The core functionality looks correct. I checked all the math in the user-facing code, and the tests seem to be checking the right stuff.

My change requests are topical in nature, and don't involve any logical changes:
- merge into this branch the docstring changes I wrote in [ba-edits/132_Update_PVSS_to_match_ISO_standard](https://github.com/MideTechnology/endaq-python/tree/ba-edits/132_Update_PVSS_to_match_ISO_standard) (main changes in f973bcf642cbf771d9a9a34480a478f60456cc03 and 13a8ef62cb628e65deb890dc309f20e9dd93dff9)
- choose a consistent naming convention for the user-facing functions. E.g., `rel_velocity` and `relative_disp_static` both abbreviate the word "relative" differently, but they *should* abbreviate it in the same way. I've commented in this review with suggestions for both an abbreviated naming convention and a non-abbreviated convention for your convenience, but feel free to use a different convention if you'd prefer.

I'll also be adding issues to address in the future the following:
- add the additional calculations to the `shock_spectrum` function as additional `mode` parameters.
- reduce redundant calculations & functions in `test_shock.py`. Atm several functions added thereto are copy-pasted versions of other functions, and could be consolidated together in some cases (notably in `*_amplitude` vs. `_phase` functions & tests) to reduce code size & the burden of maintenance.
- in the coefficient calculations, replace instances of `1 / (2Q)` with `damp` to avoid numerical errors for `damp = 0`. *(I know you expressed a desire to keep our calculations as verbatim identical to the original paper as possible, but 1) we already take some liberties in the calculation, and 2) if our tests are passing and we're still not confident that the changes were correct, then our testing isn't sufficient and we should add more comprehensive tests.)*

**Review by CrepeGoat — APPROVED** [review]
lgtm

**Pull request merged**